docs/route.md

```diff
@@ -80,42 +79,26 @@ Argument          | Type        | Required | Description
 `vnode`           | `Vnode`     | Yes      | This method is meant to be used in conjunction with an `<a>` [vnode](vnodes.md)'s [`oncreate` hook](lifecycle-methods.md)
 **returns**       | Function(e) |          | Returns an event handler that calls `m.route.set` with the link's `href` as the `path`
 
-#### RouteResolver
+#### Route components
 
-A RouterResolver is an object that contains an `onmatch` method, and optionally a `view` method.
+Since routes represent user actions and logic may need to be performed before the view can be resolved, route components can additionally, or exclusively, contain an `onmatch` method.
 
-`routeResolver = {onmatch, view}`
+##### component.onmatch
 
-##### routeResolver.onmatch
+The `onmatch` hook is called when the corresponding route is matched, before the lifecycle is engaged. It is called once when a router path changes, but not on subsequent redraws. It can be used to run logic before a component initializes (eg example authentication logic), and optionally redirect.
 
-The `onmatch` hook is called when the router needs to find a component to render. It is called once when a router path changes, but not on subsequent redraws. It can be used to run logic before a component initializes (for example authentication logic)
+This method also allows you to asynchronously assign properties to `attrs` to inform view logic, making it suitable for code splitting and asynchronous module loading.
```

Comment:
@lhorie in reply to [your comment about code splitting](https://github.com/lhorie/mithril.js/pull/1258#issuecomment-241136624) (or any kind of dynamic resolution logic), currently you would mutate the `attrs` object and let the corresponding `view` interpret it. I'd like to work at this further and provision `state` earlier, since that makes more sense — but the mechanism remains essential the same.

This makes RouteResolvers easier to reason about: a RouteResolver with an onmatch but no view *can't* resolve, and can only redirect without passing anything along; a RouteResolver with an onmatch *and* a view can either redirect to another route endpoint without passing anything on, or resolve to it's own view and have that view invoke dynamically loaded data or components or whatnot. This allows us to combine the benefits of advanced view composition (diffing the Layout component with passed in content) along with those of dynamic resource loading.

This makes the router hash more comprehensive, since you can see all view configurations from the original router (even if any number of RouteResolvers do nothing more than invoke a dynamically loaded component at the root).